### PR TITLE
LibWeb: Implement margin collapsing between parent and first child

### PR DESCRIPTION
**Pull request by kalenikaliaksandr**
Screenshot from 2022-12-23 12-19-57

### CODE REVIEW

File: Userland/Libraries/LibWeb/Layout/BlockFormattingContext.cpp [hunk 1]
```diff
@@ -361,35 +365,52 @@ void BlockFormattingContext::layout_block_level_box(Box const& box, BlockContain
...
+            } else {
+                m_margin_state.reset();
                 layout_block_level_children(verify_cast<BlockContainer>(box), layout_mode, box_state.available_inner_space_or_constraints_from(available_space));
+            }
         }
     }
 
     compute_height(box, available_space);
 
+    // FIXME: it's too naive way to check if box margins collapse through
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/Layout/BlockFormattingContext.cpp:424]: If I tried to fix this, I wouldn't know what to do. Can we add some information about what the code *should* do, maybe even with a link to the relevant spec?

File: Userland/Libraries/LibWeb/Layout/BlockFormattingContext.cpp [hunk 2]
```diff
@@ -407,12 +428,16 @@ void BlockFormattingContext::layout_block_level_children(BlockContainer const& b
...
     float bottom_of_lowest_margin_box = 0;
+    float current_y = 0;
 
     block_container.for_each_child_of_type<Box>([&](Box& box) {
-        layout_block_level_box(box, block_container, layout_mode, bottom_of_lowest_margin_box, available_space);
+        layout_block_level_box(box, block_container, layout_mode, bottom_of_lowest_margin_box, available_space, current_y);
         return IterationDecision::Continue;
     });
 
+    // FIXME: margin-bottom of last in-flow child can be collapsed with margin-bottom of parent
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/Layout/BlockFormattingContext.cpp:458]: Would be great to implement this as well in the same PR, so we see that the `BlockMarginState` concept works for all situations.

File: Userland/Libraries/LibWeb/Layout/BlockFormattingContext.h [hunk 3]
```diff
@@ -113,16 +113,36 @@ class BlockFormattingContext : public FormattingContext {
 
     struct BlockMarginState {
         Vector<float> current_collapsible_margins;
+        Optional<Function<void(float)>> update_box_waiting_for_final_y_position_callback;
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/Layout/BlockFormattingContext.h:116]: `Function` can be null, so you don't need to wrap it in `Optional` like this.

File: Userland/Libraries/LibWeb/Layout/BlockFormattingContext.h [hunk 4]
```diff
@@ -113,16 +113,36 @@ class BlockFormattingContext : public FormattingContext {
...
 
         void add_margin(float margin)
         {
             current_collapsible_margins.append(margin);
         }
 
+        void add_block_container_waiting_for_final_y_position(Function<void(float)> callback)
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/Layout/BlockFormattingContext.h:123]: The name of this API makes it sound like we're adding a `BlockContainer` to something when in reality we're registering a callback function. Let's find a name that better matches what it actually does :)

Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/Layout/BlockFormattingContext.h]: yes, makes sense. I used `block_container_y_position_update_callback` instead but probably still not perfect :)

File: Userland/Libraries/LibWeb/Layout/BlockFormattingContext.cpp (hunk 1 above)
Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/Layout/BlockFormattingContext.cpp]: makes sense, I updated this comment.

**Pull request merged**